README.md
# Rhebok

![rhebok logo](https://s3-ap-northeast-1.amazonaws.com/softwarearchives/rhebookfix-350.png)


[![Build Status](https://travis-ci.org/kazeburo/rhebok.svg?branch=master)](https://travis-ci.org/kazeburo/rhebok)

Rhebok is High Performance Rack Handler/Web Server. 2x performance when compared against Unicorn.

Rhebok supports following features.

- ultra fast HTTP processing using [picohttpparser](https://github.com/h2o/picohttpparser)
- uses accept4(2) if OS support
- uses writev(2) for output responses
- prefork and graceful shutdown using [prefork_engine](https://rubygems.org/gems/prefork_engine)
- hot deploy using [start_server](https://metacpan.org/release/Server-Starter) ([here](https://github.com/lestrrat/go-server-starter) is golang version by lestrrat-san)
- supports HTTP/1.1 except for KeepAlive
- supports OobGC

This server is suitable for running HTTP application servers behind a reverse proxy like nginx.

## Installation

Add this line to your application's Gemfile:

```
gem 'rhebok'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install rhebok

## Usage

    $ rackup -s Rhebok --port 8080 -O MaxWorkers=5 -O MaxRequestPerChild=1000 -O OobGC=yes -E production config.ru

## Sample configuration with Nginx

nginx.conf

    http {
      upstream app {
        server unix:/path/to/app.sock;
      }
      server {
        location / {
          proxy_pass http://app;
        }
        location ~ ^/(stylesheets|images)/ {
          root /path/to/webapp/public;
        }
      }
    }

command line of running Rhebok

    $ rackup -s Rhebok -O Path=/path/to/app.sock \
      -O MaxWorkers=5 -O MaxRequestPerChild=1000 -E production config.ru

## Options

### ConfigFile

filename to load options. For details, please read `Config File` section

### Host

hostname or ip address to bind (default: 0.0.0.0)

### Port

port to bind (default: 9292)

### Path

path to listen using unix socket

### BackLog

specifies a listen backlog parameter (default: Socket::SOMAXCONN. usually 128 on Linux )

### ReusePort

enable SO_REUSEPORT for TCP socket

### MaxWorkers

number of worker processes (default: 5)

### MaxRequestPerChild

Max number of requests to be handled before a worker process exits (default: 1000)
If set to `0`. worker never exists. This option looks like Apache's MaxRequestPerChild

### MinRequestPerChild

if set, randomizes the number of requests handled by a single worker process between the value and that supplied by MaxRequestPerChlid (default: none)

### Timeout

seconds until timeout (default: 300)

### OobGC

Boolean like string. If true, Rhebok execute GC after close client socket. (defualt: false)

### MaxGCPerRequest

If [gctools](https://github.com/tmm1/gctools) is available, this option is not used. invoke GC by `GC::OOB.run` after every requests.

Max number of request before invoking GC (defualt: 5)

### MinGCPerRequest

If set, randomizes the number of request before invoking GC between the number of MaxGCPerRequest (defualt: none)

### Chunked_Transfer

If set, use chunked transfer for response (default: false)

### SpawnInterval

if set, worker processes will not be spawned more than once than every given seconds. Also, when SIGUSR1 is being received, no more than one worker processes will be collected every given seconds. This feature is useful for doing a "slow-restart". See http://blog.kazuhooku.com/2011/04/web-serverstarter-parallelprefork.html for more information. (default: none)

## Config File

load options from specified config file. If same options are exists in command line and the config file, options written in the config file will take precedence.

```
$ cat rhebok_config.rb
host '127.0.0.1'
port 9292
max_workers ENV["WEB_CONCURRENCY"] || 3
before_fork {
  defined?(ActiveRecord::Base) and
    ActiveRecord::Base.connection.disconnect!
}
after_fork {
  defined?(ActiveRecord::Base) and
    ActiveRecord::Base.establish_connection
}
$ rackup -s Rhebok -O ConfigFile=rhebok_config.rb -O port 8080 -O OobGC -E production
Rhebok starts Listening on 127.0.0.1:9292 Pid:11892
```

Supported options in config file are below.

### host

### port

### path

### backlog

### reuseport

### max_workers

### timeout

### max_request_per_child

### min_request_per_child

### oobgc

### max_gc_per_request

### min_gc_per_request

### chunked_transfer

### spawn_interval

### before_fork

proc object. This block will be called by a master process before forking each worker

### after_fork

proc object. This block will be called by a worker process after forking

## Signal Handling

### Master process

- TERM, HUP: If the master process received TERM or HUP signal, Rhebok will shutdown gracefully
- USR1: If set SpawnInterval, Rhebok will collect workers every given seconds and exit

### worker process

- TERM: If the worker process received TERM, exit after finishing current request


## Benchmark

Rhebok and Unicorn "Hello World" Benchmark (behind nginx reverse proxy)

![image](https://s3-ap-northeast-1.amazonaws.com/softwarearchives/rhebok_unicorn_bench.png)

*"nginx static file" represents req/sec when delivering 13 bytes static files from nginx*

Application code is [here](https://github.com/kazeburo/rhebok_bench_app).

ruby version

    $ ruby -v
    ruby 2.1.5p273 (2014-11-13 revision 48405) [x86_64-linux]

nginx.conf

    worker_processes  16;

    events {
      worker_connections  50000;
    }

    http {
      include     mime.types;
      access_log  off;
      sendfile    on;
      tcp_nopush  on;
      tcp_nodelay on;
      etag        off;
      upstream app {
        server unix:/dev/shm/app.sock;
      }
      server {
        location / {
          proxy_pass http://app;
        }
      }
    }

### Rhebok

command to run

    $ start_server --backlog=16384 --path /dev/shm/app.sock -- \
      bundle exec --keep-file-descriptors rackup -s Rhebok \
        -O MaxWorkers=8 -O MaxRequestPerChild=0 -E production config.ru

#### Hello World/Rack Application

    $ ./wrk -t 4 -c 500 -d 30  http://localhost/
    Running 30s test @ http://localhost/
      4 threads and 500 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency     1.74ms  661.27us  38.69ms   91.19%
        Req/Sec    72.69k     9.42k  114.33k    79.43%
      8206118 requests in 30.00s, 1.34GB read
    Requests/sec: 273544.70
    Transfer/sec:     45.90MB

#### Sinatra

    $ ./wrk -t 4 -c 500 -d 30  http://localhost/
    Running 30s test @ http://localhost/
      4 threads and 500 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    16.39ms  418.08us  22.25ms   78.25%
        Req/Sec     7.73k   230.81     8.38k    70.81%
      912104 requests in 30.00s, 273.09MB read
    Requests/sec:  30404.28
    Transfer/sec:      9.10MB

#### Rails

    $ ./wrk -t 4 -c 500 -d 30  http://localhost/
    Running 30s test @ http://localhost/
      4 threads and 500 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency   101.13ms    2.57ms 139.04ms   96.88%
        Req/Sec     1.24k    27.11     1.29k    82.98%
      148169 requests in 30.00s, 178.18MB read
    Requests/sec:   4938.93
    Transfer/sec:      5.94MB

### Unicorn

unicorn.rb

    $ cat unicorn.rb
    worker_processes 8
    preload_app true
    listen "/dev/shm/app.sock"

command to run

    $ bundle exec unicorn -c unicorn.rb -E production config.ru

#### Hello World/Rack Application

    $ ./wrk -t 4 -c 500 -d 30  http://localhost/
    Running 30s test @ http://localhost/
      4 threads and 500 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency     3.50ms  518.60us  30.28ms   90.35%
        Req/Sec    37.99k     4.10k   56.56k    72.14%
      4294095 requests in 30.00s, 786.07MB read
    Requests/sec: 143140.20
    Transfer/sec:     26.20MB

#### Sinatra

    $ ./wrk -t 4 -c 500 -d 30  http://localhost/
    Running 30s test @ http://localhost/
      4 threads and 500 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    19.31ms    1.09ms  65.92ms   89.94%
        Req/Sec     6.55k   312.92     7.24k    73.41%
      775712 requests in 30.00s, 244.09MB read
    Requests/sec:  25857.85
    Transfer/sec:      8.14MB

#### Rails

    $ ./wrk -t 4 -c 500 -d 30  http://localhost/
    Running 30s test @ http://localhost/
      4 threads and 500 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency   105.70ms    4.70ms 151.11ms   93.50%
        Req/Sec     1.19k    44.16     1.25k    89.13%
      141846 requests in 30.00s, 172.74MB read
    Requests/sec:   4728.11
    Transfer/sec:      5.76MB

### Server Environment

I used EC2 for benchmarking. Instance type if c3.8xlarge(32cores). A benchmark tool and web servers were executed at same hosts. Before benchmark, increase somaxconn and nfiles.

## See Also

[Gazelle](https://metacpan.org/pod/Gazelle) Rhebok is created based on Gazelle code

[Server::Stater](https://metacpan.org/pod/Server::Starter)  a superdaemon for hot-deploying server programs

[picohttpparser](https://github.com/h2o/picohttpparser) fast http parser

## LICENSE

Copyright (C) Masahiro Nagano.

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

## Contributing

1. Fork it ( https://github.com/kazeburo/rhebok/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


Changes
0.9.3 2016-09-13T11:05:00

    - bugfix about timeout by wrong Operator Precedence #7
    - Optimize by avoiding function call #8
    - fix rack deps

0.9.2 2015-12-24T00:31:23

    - optimize a bit

0.9.1 2015-12-07T10:37:51

    - disable chunked transfer by default

0.9.0 2015-11-13T23:44:13

    - safe graceful shutdown
    - fix iovcnt calculation

0.8.6 2015-01-26T14:53:06Z

    - add some chunked transfer test. refactor around sending chunk

0.8.5 2015-01-26T02:23:08Z

    - fix bug. send 0byte mark when using a non raw Array body

0.8.3 2015-01-23T00:33:02Z

    - change default backlog to SOMAXCONN

0.8.2 2015-01-22T19:42:49Z

    - tune around Expect

0.8.1 2015-01-22T16:42:28Z

    - stringfy body and header values
    - check 0 byte body

0.8.0 2015-01-22T10:18:30Z

    - support HTTP/1.1

0.2.3 2015-01-08T11:21:01Z

    - check ENV["SERVER_STARTER_PORT"] has "="

0.2.2 2015-01-07T22:41:52Z

    - support SO_REUSEPORT

0.2.1 2014-12-27T01:26:47Z

    - fix segfault. header includes "\n"

0.2.0 2014-12-26T11:40:23Z

    - support ConfigFile
    - changed default MaxWorkers to 5

0.1.2 2014-12-25T17:33:59Z

    - oops. status_code.to_i is required

0.1.1 2014-12-25T17:19:31Z

    - added close for body

0.1.0 2014-12-25T16:23:00Z

    - remove specific rack version

0.0.8 2014-12-25T15:23:41Z

    - tweek Header order
    - require ruby >= 2.0

0.0.7 2014-12-24T16:25:22Z

    - support unix domain socket directly

0.0.6 2014-12-24T13:19:50Z

    - support OobGC

0.0.5 2014-12-19T11:06:28Z

    - support MaxRequestPerChild=0. worker never exit.
    - refactor around res header

0.0.4 2014-12-19-T01:50:11Z

    - support multiple line header

0.0.3 2014-12-18T17:22:53Z

    - fixed segfault. iovec shortage

0.0.2 2014-12-18T12:50:11Z

    - fixed install issue

0.0.1 2014-12-18T12:15:52Z

    - first release




test/testconfig.rb
host '127.0.0.1'
port 9202
max_workers 1
before_fork {
  ENV["TEST_FOO"] = "FOO"
}
after_fork {
  ENV["TEST_BAR"] = "BAR"
}


spec/rhebok_spec.rb
require 'spec_helper'

describe Rhebok do
  it 'has a version number' do
    expect(Rhebok::VERSION).not_to be nil
  end

  it 'does something useful' do
    expect(false).to eq(true)
  end
end


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'rhebok'


lib/rhebok/config.rb
class Rhebok
  class Config
    def initialize(options={})
      @config = options
    end

    def host(val)
      @config[:Host] = val
    end

    def port(val)
      @config[:Port] = val
    end

    def path(val)
      @config[:Path] = val
    end

    def max_workers(val)
      @config[:MaxWorkers] = val
    end

    def timeout(val)
      @config[:Timeout] = val
    end

    def max_request_per_child(val)
      @config[:MaxRequestPerChild] = val
    end

    def min_request_per_child(val)
      @config[:MinRequestPerChild] = val
    end

    def spawn_interval(val)
      @config[:SpawnInterval] = val
    end

    def err_respawn_interval(val)
      @config[:ErrRespawnInterval] = val
    end

    def oobgc(val)
      @config[:OobGC] = val
    end

    def max_gc_per_request(val)
      @config[:MaxGCPerRequest] = val
    end

    def min_gc_per_request(val)
      @config[:MinGCPerRequest] = val
    end

    def backlog(val)
      @config[:BackLog] = val
    end

    def before_fork(&block)
      @config[:BeforeFork] = block
    end

    def after_fork(&block)
      @config[:AfterFork] = block
    end

    def reuseport(&block)
      @config[:ReusePort] = block
    end

    def chunked_transfer(&block)
      @config[:ChunkedTransfer] = block
    end

    def retrieve
      @config
    end
  end
end


lib/rhebok/buffered.rb
require 'stringio'
require 'tempfile'

class Rhebok
  class Buffered
    def initialize(length=0,memory_max=1048576)
      @length = length
      @memory_max = memory_max
      @size = 0
      if length > memory_max
        @buffer = Tempfile.open('r')
        @buffer.binmode
        @buffer.set_encoding('BINARY')
      else
        @buffer = StringIO.new("").set_encoding('BINARY')
      end
    end

    def print(buf)
      if @size + buf.bytesize > @memory_max && @buffer.instance_of?(StringIO)
        new_buffer = Tempfile.open('r')
        new_buffer.binmode
        new_buffer.set_encoding('BINARY')
        new_buffer << @buffer.string
        @buffer = new_buffer
      end
      @buffer << buf
      @size += buf.bytesize
    end

    def size
      @size
    end

    def rewind
      @buffer.rewind
      @buffer
    end

    def close
      if @buffer.instance_of?(Tempfile)
        @buffer.close!
      end
    end
  end
end


lib/rack/handler/rhebok.rb
# -*- coding: utf-8 -*-

$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'rubygems'
require 'rack'
require 'stringio'
require 'tempfile'
require 'socket'
require 'rack/utils'
require 'io/nonblock'
require 'prefork_engine'
require 'rhebok'
require 'rhebok/config'
require 'rhebok/buffered'

$RACK_HANDLER_RHEBOK_GCTOOL = true
begin
  require 'gctools/oobgc'
rescue LoadError
  $RACK_HANDLER_RHEBOK_GCTOOL = false
end

module Rack
  module Handler
    class Rhebok
      MAX_MEMORY_BUFFER_SIZE = 1024 * 1024
      DEFAULT_OPTIONS = {
        :Host => '0.0.0.0',
        :Port => 9292,
        :Path => nil,
        :MaxWorkers => 5,
        :Timeout => 300,
        :MaxRequestPerChild => 1000,
        :MinRequestPerChild => nil,
        :SpawnInterval => nil,
        :ErrRespawnInterval => nil,
        :OobGC => false,
        :MaxGCPerRequest => 5,
        :MinGCPerRequest => nil,
        :BackLog => Socket::SOMAXCONN,
        :BeforeFork => nil,
        :AfterFork => nil,
        :ReusePort => false,
        :ChunkedTransfer => false,
      }
      NULLIO  = StringIO.new("").set_encoding('BINARY')

      def self.run(app, options={})
        slf = new(options)
        slf.setup_listener()
        slf.run_worker(app)
      end

      def self.valid_options
        {
          "Host=HOST" => "Hostname to listen on (default: 0.0.0.0)",
          "Port=PORT" => "Port to listen on (default: 9292)",
        }
      end

      def initialize(options={})
        if options[:OobGC].instance_of?(String)
          options[:OobGC] = options[:OobGC].match(/^(true|yes|1)$/i) ? true : false
        end
        if options[:ReusePort].instance_of?(String)
          options[:ReusePort] = options[:ReusePort].match(/^(true|yes|1)$/i) ? true : false
        end
        if options[:ChunkedTransfer].instance_of?(String)
          options[:ChunkedTransfer] = options[:ChunkedTransfer].match(/^(true|yes|1)$/i) ? true : false
        end

        @options = DEFAULT_OPTIONS.merge(options)
        if @options[:ConfigFile] != nil
          puts "loading from config_file:#{options[:ConfigFile]}"
          config = ::Rhebok::Config.new
          config.instance_eval(::File.read @options.delete(:ConfigFile))
          @options.merge!(config.retrieve)
        end
        @server = nil
        @_is_tcp = false
        @_using_defer_accept = false
      end

      def setup_listener()
        if ENV.has_key?("SERVER_STARTER_PORT") && ENV["SERVER_STARTER_PORT"].to_s.match(/=/)
          hostport, fd = ENV["SERVER_STARTER_PORT"].split("=",2)
          if m = hostport.match(/(.*):(\d+)/)
            @options[:Host] = m[0]
            @options[:Port] = m[1].to_i
          else
            @options[:Port] = hostport
          end
          @server = Socket.for_fd(fd.to_i)
          @_is_tcp = true if !@server.local_address.unix?
        end

        if @server == nil
          if @options[:Path] != nil
            if ::File.socket?(@options[:Path])
              puts "removing existing socket file:#{@options[:Path]}";
              ::File.unlink(@options[:Path])
            end
            begin
              ::File.unlink(@options[:Path])
            rescue
              #ignore
            end
            puts "Rhebok starts Listening on :unix:#{@options[:Path]} Pid:#{$$}"
            oldmask = ::File.umask(0)
            @server = Socket.new(Socket::AF_UNIX, Socket::SOCK_STREAM, 0)
            @server.bind(Addrinfo.unix(@options[:Path]))
            ::File.umask(oldmask)
            @_is_tcp = false
            @options[:Host] = "0.0.0.0"
            @options[:Port] = 0
          else
            puts "Rhebok starts Listening on #{@options[:Host]}:#{@options[:Port]} Pid:#{$$}"
            addrinfo = Addrinfo.tcp(@options[:Host], @options[:Port])
            @server = Socket.new(defined?(Socket::AF_INET6) && addrinfo.afamily == Socket::AF_INET6 ?
                                 Socket::AF_INET6 : Socket::AF_INET, Socket::SOCK_STREAM, 0)
            @server.setsockopt(Socket::SOL_SOCKET, Socket::SO_REUSEADDR, 1)
            if @options[:ReusePort]
              @server.setsockopt(Socket::SOL_SOCKET, Socket::SO_REUSEPORT, 1);
            end
            @server.bind(addrinfo)
            @_is_tcp = true
          end
          @server.listen(@options[:BackLog].to_i)
        end # @server == nil

        if RUBY_PLATFORM.match(/linux/) && @_is_tcp == true
          begin
            @server.setsockopt(Socket::IPPROTO_TCP, 9, 1)
            @_using_defer_accept = true
          end
        end

        if @server.respond_to?("autoclose=")
          @server.autoclose = false
        end

      end

      def run_worker(app)
        pm_args = {
          "max_workers" => @options[:MaxWorkers].to_i,
          "trap_signals" => {
            "TERM" => 'TERM',
            "HUP"  => 'TERM',
          },
        }
        if @options[:SpawnInterval]
          pm_args["trap_signals"]["USR1"] = ["TERM", @options[:SpawnInterval].to_i]
          pm_args["spawn_interval"] = @options[:SpawnInterval].to_i
        end
        if @options[:ErrRespawnInterval]
          pm_args["err_respawn_interval"] = @options[:ErrRespawnInterval].to_i
        end
        if @options[:BeforeFork]
          pm_args["before_fork"] = proc { |pe2|
            @options[:BeforeFork].call
          }
        end
        Signal.trap('INT','SYSTEM_DEFAULT') # XXX

        pe = PreforkEngine.new(pm_args)
        while !pe.signal_received.match(/^(TERM|USR1)$/)
          pe.start do
            srand
            if @options[:AfterFork]
              @options[:AfterFork].call
            end
            self.accept_loop(app)
          end
        end

        while pe.wait_all_children(1) > 0
          pe.signal_all_children('TERM')
        end
      end

      def _calc_reqs_per_child
        if @options[:MinRequestPerChild] == nil
          return @options[:MaxRequestPerChild].to_i
        end
        max = @options[:MaxRequestPerChild].to_i
        min = @options[:MinRequestPerChild].to_i
        if min < max
          max - ((max - min + 1) * rand).to_i
        else
          max
        end
      end

      def _calc_gc_per_req
        if @options[:MinGCPerRequest] == nil
          return @options[:MaxGCPerRequest].to_i
        end
        max = @options[:MaxGCPerRequest].to_i
        min = @options[:MinGCPerRequest].to_i
        if min < max
          max - ((max - min + 1) * rand).to_i
        else
          max
        end
      end


      def accept_loop(app)
        @term_received = 0
        proc_req_count = 0
        Signal.trap(:TERM) do
          @term_received += 1
        end
        Signal.trap(:PIPE, "IGNORE")
        max_reqs = self._calc_reqs_per_child()
        gc_reqs = self._calc_gc_per_req()
        fileno = @server.fileno

        env_template = {
          "SERVER_NAME"       => @options[:Host],
          "SERVER_PORT"       => @options[:Port].to_s,
          "rack.version"      => [1,1],
          "rack.errors"       => STDERR,
          "rack.multithread"  => false,
          "rack.multiprocess" => true,
          "rack.run_once"     => false,
          "rack.url_scheme"   => "http",
          "rack.input"        => NULLIO
        }

        while @options[:MaxRequestPerChild].to_i == 0 || proc_req_count < max_reqs
          if @term_received > 0
            exit!(true)
          end
          env = env_template.clone
          connection, buf = ::Rhebok.accept_rack(fileno, @options[:Timeout], @_is_tcp, env)
          if connection
            # for tempfile
            buffer = nil
            begin
              proc_req_count += 1
              # handle request
              if env.key?("CONTENT_LENGTH") && env["CONTENT_LENGTH"].to_i > 0
                cl = env["CONTENT_LENGTH"].to_i
                buffer = ::Rhebok::Buffered.new(cl,MAX_MEMORY_BUFFER_SIZE)
                while cl > 0
                  chunk = ""
                  if buf.bytesize > 0
                    chunk = buf
                    buf = ""
                  else
                    readed = ::Rhebok.read_timeout(connection, chunk, cl, 0, @options[:Timeout])
                    if readed == nil
                      return
                    end
                  end
                  buffer.print(chunk)
                  cl -= chunk.bytesize
                end
                env["rack.input"] = buffer.rewind
              elsif env.key?("HTTP_TRANSFER_ENCODING") && env.delete("HTTP_TRANSFER_ENCODING") == 'chunked'
                buffer = ::Rhebok::Buffered.new(0,MAX_MEMORY_BUFFER_SIZE)
                chunked_buffer = '';
                complete = false
                while !complete
                  chunk = ""
                  if buf.bytesize > 0
                    chunk = buf
                    buf = ""
                  else
                    readed = ::Rhebok.read_timeout(connection, chunk, 16384, 0, @options[:Timeout])
                    if readed == nil
                      return
                    end
                  end
                  chunked_buffer << chunk
                  while chunked_buffer.sub!(/^(([0-9a-fA-F]+).*\015\012)/,"") != nil
                    trailer = $1
                    chunked_len = $2.hex
                    if chunked_len == 0
                      complete = true
                      break
                    elsif chunked_buffer.bytesize < chunked_len + 2
                      chunked_buffer = trailer + chunked_buffer
                      break
                    end
                    buffer.print(chunked_buffer.byteslice(0,chunked_len))
                    chunked_buffer = chunked_buffer.byteslice(chunked_len,chunked_buffer.bytesize-chunked_len)
                    chunked_buffer.sub!(/^\015\012/,"")
                  end
                  break if complete
                end
                env["CONTENT_LENGTH"] = buffer.size.to_s
                env["rack.input"] = buffer.rewind
              end

              status_code, headers, body = app.call(env)

              use_chunked = 0
              if @options[:ChunkedTransfer]
                use_chunked =  env["SERVER_PROTOCOL"] != "HTTP/1.1" ||
                               headers.key?("Transfer-Encoding") ||
                               headers.key?("Content-Length") ? 0 : 1
              end

              if body.instance_of?(Array)
                ::Rhebok.write_response(connection, @options[:Timeout], status_code.to_i, headers, body, use_chunked, 0)
              else
                ::Rhebok.write_response(connection, @options[:Timeout], status_code.to_i, headers, [], use_chunked, 1)
                body.each do |part|
                  ret = nil
                  if use_chunked == 1
                    ret = ::Rhebok.write_chunk(connection, part, 0, @options[:Timeout])
                  else
                    ret = ::Rhebok.write_all(connection, part, 0, @options[:Timeout])
                  end
                  if ret == nil
                    break
                  end
                end #body.each
                ::Rhebok.write_all(connection, "0\015\012\015\012", 0, @options[:Timeout]) if use_chunked == 1
                body.respond_to?(:close) and body.close
              end
              #p [env,status_code,headers,body]
            ensure
              if buffer != nil
                buffer.close
              end
              ::Rhebok.close_rack(connection)
              # out of band gc
              if @options[:OobGC]
                if $RACK_HANDLER_RHEBOK_GCTOOL
                  GC::OOB.run
                elsif proc_req_count % gc_reqs == 0
                  disabled = GC.enable
                  GC.start
                  GC.disable if disabled
                end
              end
            end #begin
          end # accept
        end #while max_reqs
      end #def

    end
  end
end
